### PR TITLE
[refactor] fix warings when compile with clang

### LINKED ISSUE #8013: [Enhancement] Open clang compile warning option and resolve compile error
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Description

There are still these warnings left to be resolved:
```cmake
        -Wno-cast-align
        -Wno-cast-qual
        -Wno-comma
        -Wno-conditional-uninitialized
        -Wno-constant-conversion
        -Wno-covered-switch-default
        -Wno-delete-abstract-non-virtual-dtor
        -Wno-delete-non-abstract-non-virtual-dtor
        -Wno-deprecated
        -Wno-deprecated-copy-with-dtor
        -Wno-deprecated-copy-with-user-provided-copy
        -Wno-deprecated-copy-with-user-provided-dtor
        -Wno-deprecated-declarations
        -Wno-deprecated-this-capture
        -Wno-documentation
        -Wno-embedded-directive
        -Wno-enum-compare-switch
        -Wno-extra-semi
        -Wno-extra-semi-stmt
        -Wno-format-nonliteral
        -Wno-frame-larger-than
        -Wno-gnu-label-as-value
        -Wno-gnu-redeclared-enum
        -Wno-gnu-statement-expression
        -Wno-gnu-zero-variadic-macro-arguments
        -Wno-header-hygiene
        -Wno-ignored-qualifiers
        -Wno-implicit-const-int-float-conversion
        -Wno-implicit-exception-spec-mismatch
        -Wno-implicit-fallthrough
        -Wno-inconsistent-missing-destructor-override
        -Wno-inconsistent-missing-override
        -Wno-instantiation-after-specialization
        -Wno-macro-redefined
        -Wno-mismatched-new-delete
        -Wno-mismatched-tags
        -Wno-missing-field-initializers
        -Wno-missing-noreturn
        -Wno-newline-eof
        -Wno-non-virtual-dtor
        -Wno-null-conversion
        -Wno-null-dereference
        -Wno-old-style-cast
        -Wno-overloaded-virtual
        -Wno-pedantic
        -Wno-pessimizing-move
        -Wno-pragma-once-outside-header
        -Wno-reserved-identifier
        -Wno-reserved-macro-identifier
        -Wno-return-type
        -Wno-shadow
        -Wno-shadow-field
        -Wno-shadow-field-in-constructor
        -Wno-shadow-uncaptured-local
        -Wno-shift-count-overflow
        -Wno-sign-compare
        -Wno-string-plus-int
        -Wno-suggest-destructor-override
        -Wno-suggest-override
        -Wno-tautological-constant-out-of-range-compare
        -Wno-tautological-unsigned-zero-compare
        -Wno-undef
        -Wno-uninitialized
        -Wno-unknown-pragmas
        -Wno-unknown-warning-option
        -Wno-unreachable-code
        -Wno-unreachable-code-break
        -Wno-unreachable-code-return
        -Wno-unused-lambda-capture
        -Wno-unused-macros
        -Wno-unused-member-function
        -Wno-unused-parameter
        -Wno-unused-private-field
        -Wno-unused-result
```

### Solution

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by yangzhg**
# Proposed changes

Issue Number: close #8013 

## Problem Summary:

fix warings when compile with clang

## Checklist(Required)

1. Does it affect the original behavior: (No)
2. Has unit tests been added: (No)
3. Has document been added or modified: (No Need)
4. Does it need to update dependencies: (No)
5. Are there any changes that cannot be rolled back: (Yes)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Review by morningman — APPROVED** [review]
Although there are some minor problems, it is still Approve first to prevent subsequent conflicts.
These issues can be fixed in subsequent PRs.

File: be/src/util/arrow/row_batch.cpp [hunk 1]
```diff
@@ -300,7 +300,7 @@ class FromRowBatchConverter : public arrow::TypeVisitor {
         arrow::NumericBuilder<T> builder(_pool);
 
         size_t num_rows = _batch.num_rows();
-        builder.Reserve(num_rows);
+        ARROW_RETURN_NOT_OK(builder.Reserve(num_rows));
```
[the remark this answers is not in the document]
Comment (yangzhg) [reply be/src/util/arrow/row_batch.cpp:303]: yes, Reserve is not void function, we do not use the return value

File: be/src/vec/data_types/data_type_decimal.h [hunk 2]
```diff
@@ -330,19 +330,9 @@ convert_to_decimal(const typename FromDataType::FieldType& value, UInt32 scale)
...
-                LOG(FATAL) << "Decimal convert overflow. Float is out of Decimal range";
-            }
-        } else {
-            if (out <= std::numeric_limits<ToNativeType>::min() ||
-                out >= std::numeric_limits<ToNativeType>::max()) {
-                LOG(FATAL) << "Decimal convert overflow. Float is out of Decimal range";
-            }
+        if (out <= static_cast<FromFieldType>(std::numeric_limits<ToNativeType>::min()) ||
+            out >= static_cast<FromFieldType>(std::numeric_limits<ToNativeType>::max())) {
+            LOG(FATAL) << "Decimal convert overflow. Float is out of Decimal range";
```
[the remark this answers is not in the document]
Comment (yangzhg) [reply be/src/vec/data_types/data_type_decimal.h:335]: numeric_limits can handle int128 min and max

File: be/src/vec/functions/function_cast.h [hunk 3]
```diff
@@ -374,9 +374,9 @@ struct ToNumberMonotonicity {
             Float64 right_float = right.get<Float64>();
 
             if (left_float >= std::numeric_limits<T>::min() &&
-                left_float <= std::numeric_limits<T>::max() &&
+                left_float <= static_cast<Float64>(std::numeric_limits<T>::max()) &&
```
[the remark this answers is not in the document]
Comment (yangzhg) [reply be/src/vec/functions/function_cast.h:377]: min is safe to cast to Float64

**Pull request merged**